crucible/test/helpers/SymSequence.hs

```diff
@@ -0,0 +1,297 @@
+{-# LANGUAGE EmptyDataDeriving #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module SymSequence (tests) where
+
+import Control.Monad.IO.Class (liftIO)
+import Data.Foldable qualified as F
+import Data.List qualified as List
+import Data.Maybe qualified as Maybe
+import Data.Parameterized.Nonce qualified as Nonce
+import Data.Parameterized.Some (Some(Some))
+import Hedgehog (Gen)
+import Hedgehog qualified as H
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Range qualified as Range
+import Lang.Crucible.Backend (SomeBackend(SomeBackend), backendGetSym)
+import Lang.Crucible.Backend.Simple (newSimpleBackend)
+import Lang.Crucible.Simulator.SymSequence (SymSequence)
+import Lang.Crucible.Simulator.SymSequence qualified as S
+import Test.Tasty qualified as TT
+import Test.Tasty.Hedgehog qualified as TTH
+import What4.Expr (EmptyExprBuilderState(EmptyExprBuilderState))
+import What4.Expr.Builder (newExprBuilder)
+import What4.FloatMode (FloatModeRepr(FloatIEEERepr))
+import What4.Interface qualified as WI
+import What4.Partial qualified as WP
+
+---------------------------------------------------------------------
+-- Tests
+
+tests :: TT.TestTree
+tests =
+  TTH.testProperty
+    "propSame"
+    -- This is a big API, so we want adequate coverage (default is 100)
+    (H.withTests 4096 propSame)
+
+propSame :: H.Property
+propSame =
+  H.property $ do
+    Some (SomeBackend bak) <- liftIO mkBackend
+    let sym = backendGetSym bak
+    op <- H.forAll (Gen.sized $ \n -> genList (H.unSize n) Gen.bool)
+    let l = opList op
+    s <- liftIO (opSeq sym op)
+    l' <- liftIO (F.toList <$> asSeq sym s)
+    l H.=== l'
+  where
+    asSeq sym =
+      S.concretizeSymSequence (pure . asConstPred (Just sym)) pure
+
+---------------------------------------------------------------------
+-- Helpers
+
+mkBackend :: IO (Some SomeBackend)
+mkBackend = do
+  sym <- newExprBuilder FloatIEEERepr EmptyExprBuilderState Nonce.globalNonceGenerator
+  Some . SomeBackend <$> newSimpleBackend sym
+
+asConstPred ::
```

Comment:
Worth documenting that this has a precondition that the supplied `Pred` is concrete.